### ISSUE TITLE
Low entropy after enclave boot (reading from /dev/random blocks)

### ISSUE THREAD

**Issue by ppmag** [aws/aws-nitro-enclaves-sdk-c#41]
Hi,

I just faced the problem, a brand new enclave has zero entropy:

```
cat /proc/sys/kernel/random/entropy_avail
0
```

which causing block when reading from /dev/random early after boot.

What about adding some entropy from NSM RNG at enclave boot stage ?

(in `nitro-cli run-enclave`? )
(or at lower level:  I just submitted issue there - https://github.com/aws/aws-nitro-enclaves-sdk-bootstrap/issues/5)

Another question: which preferred workaround to use right now?

**Comment by petreeftime** [comment]
There a few possible solutions: 
1. Call [`aws_nitro_enclaves_library_seed_entropy`](https://github.com/aws/aws-nitro-enclaves-sdk-c/blob/main/include/aws/nitro_enclaves/nitro_enclaves.h#L34) explicitly during application start, and on a timer afterwards. 
2. Use RDSEED CPU instruction to seed entropy, [`rngd`](https://github.com/nhorman/rng-tools) or similar can help with this.
3. Set `random.trust_cpu=on` in the kernel commandline (only works on Linux kernels > 4.19).

It might be possible to integrate calls to generate entropy and seed entropy into the provided init process of the enclave, but might not be the best option.

**Comment by ppmag** [comment]
Thank you very much,
first option (`aws_nitro_enclaves_library_seed_entropy()`) works like a charm!

**Comment by Jonas-Metzger** [comment]
Hi @petreeftime , is there an easy way to call `aws_nitro_enclaves_library_seed_entropy` from the command line?

**Comment by NullHypothesis** [comment]
I attempted to use solution 1 to supply the system with entropy but it doesn't have the expected effect. Here's what I did:

1. Run a simple executable that calls `aws_nitro_enclaves_library_seed_entropy(1024)` periodically.
2. Look at `/proc/sys/kernel/random/entropy_avail`, which is still 0.
3. Start a separate application that reads from `/dev/urandom`; the read call blocks.

What am I missing here? For what it's worth, I'm running my enclave in debug mode (in case that makes a difference) and here are the last four log messages from startup:
```
[    1.024374] nsm: loading out-of-tree module taints kernel.
[    1.025030] nsm: module verification failed: signature and/or required key missing - tainting kernel
[    1.033770] random: entropy-init: uninitialized urandom read (32 bytes read)
[    1.034594] random: entropy-init: uninitialized urandom read (32 bytes read)
```
